endtoend/components/validator.go

```diff
@@ -19,18 +19,164 @@ import (
+			case <-node.Started():
+				continue
+			}
+		}
+		// Only close if all nodes are started. When context is cancelled, node is not considered as started.
+		// Stalled nodes do not pose problem -- after certain period of time main routine times out.
+		close(s.started)
+	}()
+
+	return g.Wait()
+}
+
+// Started checks whether validator node set is started and all nodes are ready to be queried.
+func (s *ValidatorNodes) Started() <-chan struct{} {
+	return s.started
+}
+
+// ValidatorNode represents a validator node.
+type ValidatorNode struct {
+	e2etypes.ComponentRunner
+	config       *e2etypes.E2EConfig
+	started      chan struct{}
+	validatorNum int
+	index        int
+	offset       int
+}
+
+// NewValidatorNode creates and returns a validator node.
+func NewValidatorNode(config *e2etypes.E2EConfig, validatorNum, index, offset int) *ValidatorNode {
+	return &ValidatorNode{
+		config:       config,
+		validatorNum: validatorNum,
+		index:        index,
+		offset:       offset,
+		started:      make(chan struct{}, 1),
+	}
+}
+
+// StartNewValidatorNode starts a validator client.
+func (v *ValidatorNode) Start(ctx context.Context) error {
+	binaryPath, found := bazel.FindBinary("cmd/validator", "validator")
+	if !found {
+		return errors.New("validator binary not found")
+	}
+
+	config, validatorNum, index, offset := v.config, v.validatorNum, v.index, v.offset
+	beaconRPCPort := e2e.TestParams.BeaconNodeRPCPort + index
+	if beaconRPCPort >= e2e.TestParams.BeaconNodeRPCPort+e2e.TestParams.BeaconNodeCount {
+		// Point any extra validator clients to a node we know is running.
+		beaconRPCPort = e2e.TestParams.BeaconNodeRPCPort
+	}
+
+	file, err := helpers.DeleteAndCreateFile(e2e.TestParams.LogPath, fmt.Sprintf(e2e.ValidatorLogFileName, index))
+	if err != nil {
+		return err
+	}
+	gFile, err := helpers.GraffitiYamlFile(e2e.TestParams.TestPath)
+	if err != nil {
+		return err
+	}
+	args := []string{
+		fmt.Sprintf("--datadir=%s/eth2-val-%d", e2e.TestParams.TestPath, index),
+		fmt.Sprintf("--log-file=%s", file.Name()),
+		fmt.Sprintf("--graffiti-file=%s", gFile),
+		fmt.Sprintf("--interop-num-validators=%d", validatorNum),
+		fmt.Sprintf("--interop-start-index=%d", offset),
+		fmt.Sprintf("--monitoring-port=%d", e2e.TestParams.ValidatorMetricsPort+index),
+		fmt.Sprintf("--grpc-gateway-port=%d", e2e.TestParams.ValidatorGatewayPort+index),
+		fmt.Sprintf("--beacon-rpc-provider=localhost:%d", beaconRPCPort),
+		"--grpc-headers=dummy=value,foo=bar", // Sending random headers shouldn't break anything.
+		"--force-clear-db",
+		"--e2e-config",
+		"--accept-terms-of-use",
+		"--verbosity=debug",
+	}
+	args = append(args, featureconfig.E2EValidatorFlags...)
+	args = append(args, config.ValidatorFlags...)
+
+	cmd := exec.CommandContext(ctx, binaryPath, args...)
+	log.Infof("Starting validator client %d with flags: %s", index, strings.Join(args[2:], " "))
+	if err = cmd.Start(); err != nil {
+		return err
+	}
+
+	// Mark node as ready.
+	close(v.started)
```

Comment:
Using `struct{}` instead of `bool` (in order not to allocate memory, is as idiomatic as it gets. Closing channel (when you have a single writer and potentially many readers) to signal sth is idiomatic as well -- no point to send dummy value, if the only fact we are interested in is that sth is ready. Overall, I think that it is a well known pattern, and is simple to implement debug.